File: app/src/main/java/com/task/ui/component/news/HomeActivity.java
package com.task.ui.component.news;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.VisibleForTesting;
import com.google.android.material.snackbar.Snackbar;
import androidx.test.espresso.IdlingResource;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.task.App;
import com.task.R;
import com.task.data.remote.dto.NewsItem;
import com.task.ui.base.BaseActivity;
import com.task.ui.component.details.DetailsActivity;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;

import static com.google.android.material.snackbar.Snackbar.LENGTH_SHORT;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.task.utils.Constants.NEWS_ITEM_KEY;
import static com.task.utils.EspressoIdlingResource.decrement;
import static com.task.utils.EspressoIdlingResource.getIdlingResource;
import static com.task.utils.EspressoIdlingResource.increment;

/**
 * Created by AhmedEltaher on 5/12/2016
 */

public class HomeActivity extends BaseActivity implements HomeContract.View {
    @Inject
    HomePresenter presenter;
    @BindView(R.id.rv_news_list)
    RecyclerView rvNews;
    @BindView(R.id.pb_loading)
    ProgressBar pbLoading;
    @BindView(R.id.tv_no_data)
    TextView tvNoData;
    @BindView(R.id.rl_news_list)
    RelativeLayout rlNewsList;
    @BindView(R.id.btn_search)
    ImageButton btnSearch;
    @BindView(R.id.et_search)
    EditText editTextSearch;

    @Override
    protected void initializeDagger() {
        App app = (App) getApplicationContext();
        app.getMainComponent().inject(HomeActivity.this);
    }

    @Override
    protected void initializePresenter() {
        super.presenter = presenter;
        presenter.setView(this);
    }

    @Override
    public int getLayoutId() {
        return R.layout.home_activity;
    }

    @Override
    public void initializeNewsList(List<NewsItem> news) {
        NewsAdapter newsAdapter = new NewsAdapter(presenter.getRecyclerItemListener(), news);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        rvNews.setLayoutManager(layoutManager);
        rvNews.setHasFixedSize(true);
        rvNews.setAdapter(newsAdapter);
    }

    @Override
    public void setLoaderVisibility(boolean isVisible) {
        pbLoading.setVisibility(isVisible ? VISIBLE : GONE);
    }

    @Override
    public void navigateToDetailsScreen(NewsItem news) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(NEWS_ITEM_KEY, news);
        Intent intent = new Intent(this, DetailsActivity.class);
        intent.putExtras(bundle);
        startActivity(intent);
    }

    @Override
    public void setNoDataVisibility(boolean isVisible) {
        tvNoData.setVisibility(isVisible ? VISIBLE : GONE);
    }

    @Override
    public void setListVisibility(boolean isVisible) {
        rlNewsList.setVisibility(isVisible ? VISIBLE : GONE);
    }

    @Override
    public void showSearchError() {
        Snackbar.make(rlNewsList, getString(R.string.search_error), LENGTH_SHORT).show();
    }

    @Override
    public void showNoNewsError() {
        Snackbar.make(rlNewsList, getString(R.string.news_error), LENGTH_SHORT).show();
    }

    @Override
    public void incrementCountingIdlingResource() {
        increment();
    }

    @Override
    public void decrementCountingIdlingResource() {
        decrement();
    }

    @OnClick({R.id.ic_toolbar_setting, R.id.ic_toolbar_refresh, R.id.btn_search})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.ic_toolbar_refresh:
                presenter.getNews();
                break;
            case R.id.btn_search:
                presenter.onSearchClick(editTextSearch.getText().toString());
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.unSubscribe();
    }

    @VisibleForTesting
    public IdlingResource getCountingIdlingResource() {
        return getIdlingResource();
    }
}

File: app/src/main/java/com/task/ui/component/news/HomeActivity.kt
package com.task.ui.component.news

import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import androidx.annotation.VisibleForTesting
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.test.espresso.IdlingResource
import butterknife.OnClick
import com.task.App
import com.task.R
import com.task.data.remote.dto.NewsItem
import com.task.ui.base.BaseActivity
import com.task.ui.component.details.DetailsActivity
import com.task.utils.Constants
import com.task.utils.EspressoIdlingResource
import kotlinx.android.synthetic.main.home_activity.*
import org.jetbrains.anko.design.snackbar
import org.jetbrains.anko.intentFor
import javax.inject.Inject

/**
 * Created by AhmedEltaher on 5/12/2016
 */

class HomeActivity : BaseActivity(), HomeContract.View {
    @Inject
    lateinit var homePresenter: HomePresenter

    override val layoutId: Int
        get() = R.layout.home_activity

    val countingIdlingResource: IdlingResource
        @VisibleForTesting
        get() = EspressoIdlingResource.idlingResource

    override fun initializeDagger() {
        val app = applicationContext as App
        app.mainComponent?.inject(this@HomeActivity)
    }

    override fun initializePresenter() {
        homePresenter.setView(this)
        super.presenter = homePresenter

    }

    override fun initializeNewsList(news: List<NewsItem>) {
        val newsAdapter = NewsAdapter(homePresenter.getRecyclerItemListener(), news)
        val layoutManager = LinearLayoutManager(this)
        rv_news_list.layoutManager = layoutManager
        rv_news_list.setHasFixedSize(true)
        rv_news_list.adapter = newsAdapter
    }

    override fun setLoaderVisibility(isVisible: Boolean) {
        pb_loading.visibility = if (isVisible) VISIBLE else GONE
    }

    override fun navigateToDetailsScreen(news: NewsItem) {
        startActivity(intentFor<DetailsActivity>(Constants.NEWS_ITEM_KEY to news))
    }

    override fun setNoDataVisibility(isVisible: Boolean) {
        tv_no_data.visibility = if (isVisible) VISIBLE else GONE
    }

    override fun setListVisibility(isVisible: Boolean) {
        rl_news_list.visibility = if (isVisible) VISIBLE else GONE
    }

    override fun showSearchError() {
        rl_news_list.snackbar(R.string.search_error)
    }

    override fun showNoNewsError() {
        rl_news_list.snackbar(R.string.news_error)
    }

    override fun incrementCountingIdlingResource() {
        EspressoIdlingResource.increment()
    }

    override fun decrementCountingIdlingResource() {
        EspressoIdlingResource.decrement()
    }

    @OnClick(R.id.ic_toolbar_setting, R.id.ic_toolbar_refresh, R.id.btn_search)
    fun onClick(view: View) {
        when (view.id) {
            R.id.ic_toolbar_refresh -> homePresenter.getNews()
            R.id.btn_search -> homePresenter.onSearchClick(et_search.text.toString())
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        homePresenter.unSubscribe()
    }
}
